### PR TITLE
General improvements and fixes

### PR DESCRIPTION
**Pull request by tecosaur**
I'm doing a big batch of changes to tighten up the current code and test it better, making a few fixes along the way.

### CODE REVIEW

**Comment by tecosaur** [comment]
Progress update: I've added ~100 new tests over ~400 lines, and written a ~200 line fuzzer using `Supposition` with the help of @Seelengrab. 4 minor bugs have been fixed, one more major one (off by 1 error and unsigned integer overflow in the 8-bit colour approximation function).

All in all, I'm very happy with how this sprint of work is shaping up.

**Comment by tecosaur** [comment]
I think this round of polishing might be approaching "done". If anybody has thoughts for more aspects that could do with attention, do say so here soon.

File: src/io.jl [hunk 1]
```diff
@@ -54,24 +54,31 @@ Print to `io` the best 8-bit SGR color code that sets the `category` color to
...
+    to6cube(value) = (value - 35) ÷ 40
+    from6cube(r6, g6, b6) = 16 + 6^2 * r6 + 6^1 * g6 + 6^0 * b6
+    sixcube = (0, 95:40:255...)
     r6cube, g6cube, b6cube = to6cube(r), to6cube(g), to6cube(b)
-    sixcube = (0, 95, 135, 175, 215, 255)
-    rnear, gnear, bnear = sixcube[r6cube], sixcube[g6cube], sixcube[b6cube]
+    rnear, gnear, bnear = sixcube[r6cube+1], sixcube[g6cube+1], sixcube[b6cube+1]
     colorcode = if r == rnear && g == gnear && b == bnear
-        16 + 35 * r6cube + 6 * g6cube + b6cube
+        from6cube(r6cube, g6cube, b6cube)
```
Comment (caleb-allen) [inline src/io.jl:70]: Bless you for these comments, this stuff is wild

File: test/runtests.jl [hunk 2]
```diff
@@ -1,8 +1,29 @@
...
+    @styled_str, styled, StyledMarkup, getface, addface!, loadface!, resetfaces!
+using .StyledMarkup: MalformedStylingMacro
+using Base: AnnotatedString, AnnotatedChar, AnnotatedIOBuffer, annotations
+
+# For output testing
+
+const kitty = Base.TermInfo(read(joinpath(@__DIR__, "terminfos", "kitty"), Base.TermInfoRaw))
+const vt100 = Base.TermInfo(read(joinpath(@__DIR__, "terminfos", "vt100"), Base.TermInfoRaw))
+
+function with_terminfo(fn::Function, tinfo::Base.TermInfo)
```
Comment (caleb-allen) [inline test/runtests.jl:15]: Is the terminfo license MIT compatible? Would want to be certain before distributing this as part of the stdlib.

**Review by caleb-allen — COMMENTED** [review]
Just one license related question, but this looks great!

File: test/runtests.jl (hunk 2 above)
Comment (tecosaur) [reply test/runtests.jl:15]: Hmm, good point.

From what I read, it looks like the vt100 file will follow the general ncurses database license, but the kitty terminfo file is originally sourced from https://github.com/kovidgoyal/kitty which is GPL-3. There's also the aspect where it look like it may be a stretch whether compiled terminfo files themselves are copyrightable, since they are a machine-readable description of the method of operation for the terminal, with little to no room for variation in expression.

In perhaps a similar vein, I think it's worth noting that recipes are not copyrightable, with a comment from a [US government FAQ](https://www.copyright.gov/circs/circ33.pdf) on this reading

> A recipe is a statement of the ingredients and procedure required for making a dish of food. A mere listing of ingredients or contents, or a simple set of directions, is uncopyrightable. As a result, the Office cannot register recipes consisting of a set of ingredients and a process for preparing a dish.

Given a terminfo file is a statement of recognised ANSI codes and their forms, I wouldn't be surprised if somebody more legally knowledgeable than me were to say the logic applied to a list of ingredients in a recipe also applies to a list of Termcap names and their ANSI codes.

Comment (caleb-allen) [reply test/runtests.jl:15]: I agree with you in principle, that it appears difficult to describe these binaries as a work that falls under copyright, but distributing something which originates in any form from a GPL-3 licensed codebase spooks me. 

I'm not a lawyer and I can't say that your interpretation is wrong. Still, without advice from an actual lawyer, it may be prudent to avoid the risk entirely by removing the files from the source. The fact that its copyright status is not immediately apparent is enough to warrant extreme caution, in my opinion.

Is it possible to generate the terminfos files as a part of the testing script? Or to fetch them from the original distributors? In this way, the files can still be used without the risk of distributing something that may or may not be licensed under GPL-3.

File: test/fuzz.jl [hunk 3]
```diff
@@ -0,0 +1,178 @@
+using StyledStrings: styled
+using Supposition
```
Comment (mbauman) [inline test/fuzz.jl:2]: Add Supposition into the project's `[extras]` as a test target?

Comment (tecosaur) [reply test/fuzz.jl:2]: I could, and do the same trick as Tar.jl does to only run the test in CI, however, it's not terribly quick to run (~30s), and only relevant when changes to `styledmarkup.jl` are made. This is why I've left it as just a big note of "run this if you edit this file" at the start of `styledmarkup.jl` for now.

I'm not sure what the best course of action is, this is just what's occurred to me.

Comment (Seelengrab) [reply test/fuzz.jl:2]: You can cut the execution time of a "everything passes" invocation of `@check` that's more-or-less only there to safeguard against regressions by setting a smaller  `max_examples`, from the default of `10_000`. E.g. if you set `max_examples = 1_000`, that should roughly cut execution time by a factor of ten. This shouldn't impact the quality of counterexamples that are found, since `max_examples` is just for steering the maximum of examples before Supposition.jl "gives up" and claims that the property holds.

Comment (tecosaur) [reply test/fuzz.jl:2]: Right, I wonder if that might make it easier for corner cases to slip in if people thinks they can trust CI and don't run it locally with more examples.

Comment (Seelengrab) [reply test/fuzz.jl:2]: Rare corner cases can always happen :shrug: Supposition.jl is not generally exhaustive, but neither are hard-coded example tests. The former at least has the chance of uncovering it eventually (and reproducibly at that), while the latter gives the illusion of correctness.

I should look into making it easier to get a potential failure out of CI and onto a local machine though, maybe by having an example CI job that stores the used `ExampleDB` somewhere other than the local filesystem or having e.g. an S3-backed `ExampledDB`? That's not something you have to worry about for your PR though.

File: test/runtests.jl (hunk 2 above)
Comment (tecosaur) [reply test/runtests.jl:15]: I've just switched to the wezterm terminfo, which should have all the capabilities we test for but comes from a MIT-licenced codebase.

Comment (tecosaur) [reply test/runtests.jl:15]: Ah darn it, it fails to set the `setrgbf` parameter.

Comment (tecosaur) [reply test/runtests.jl:15]: Screw it, I'll make a custom terminfo.

File: src/io.jl (hunk 1 above)
Comment (tecosaur) [reply src/io.jl:70]: It feels a lot better to have cleaner code to look at too :sweat_smile:

File: test/fuzz.jl (hunk 3 above)
Comment (tecosaur) [reply test/fuzz.jl:5]: We now have fuzz-testing as part of CI :slightly_smiling_face:

Comment (tecosaur) [reply test/fuzz.jl:5]: Looks like I'll need to adjust the tests/CI to work as expected (specifically `hasparserdiverged()`)

Comment (tecosaur) [reply test/fuzz.jl:5]: Ah, here we go
```
Test Summary: | Pass  Total   Time
Fuzz (5000)   |    1      1  28.8s
```

Should do `Fuzz (10000)` locally, and `Fuzz (1000)` on main or PRs that don't change `styledmarkup.jl`.

File: docs/src/examples.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
+# Usage Examples
+
+!!! info
+    There are some examples on this page that don't quite look as they should see [Documenter.jl#2488](https://github.com/JuliaDocs/Documenter.jl/issues/2488) for more information and potential improvements to this situation.
```
Comment (lazarusA) [inline docs/src/examples.md:4]:
```suggestion
    There are some examples on this page that don't quite look as they should, see [Documenter.jl#2488](https://github.com/JuliaDocs/Documenter.jl/issues/2488) for more information and potential improvements to this situation.
```

File: docs/src/examples.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+printer knows to handle the ANSI named colours specially, but you can create
+more "named colours" simply by adding new faces.
+
+```@repl examples
+StyledStrings.loadface!(:orange => StyledStrings.Face(foreground = 0xFF7700))
+styled"{orange:this is orange text}"
+```
+
+The fact that named colors are implemented this way also allows for other nice
+conveniences. For example, if you waned a more subtle version of the warning
```
Comment (lazarusA) [inline docs/src/examples.md:46]:
```suggestion
conveniences. For example, if you want a more subtle version of the warning
```

File: docs/src/examples.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+```
+
+Should you want to use a particular colour just once, the `foreground`/`fg` and `background`/`bg` inline face attributes can be set to hex codes.
+
+```@repl examples
+styled"{(fg=#4063d8):ju}{(fg=#389826):l}{(fg=#cb3c33):i}{(fg=#9558b2):a}"
+```
+
+It is recommended that package authors create faces with a focus on the semantic
+meaning they wish to impart, and then consider what styling suits. For example,
```
Comment (lazarusA) [inline docs/src/examples.md:62]: impart? or import

File: docs/src/examples.md (hunk 5 above)
Comment (tecosaur) [reply docs/src/examples.md:46]: Ah yep, thanks

File: docs/src/examples.md (hunk 4 above)
Comment (tecosaur) [reply docs/src/examples.md:4]: That comma helps :)

File: docs/src/examples.md (hunk 6 above)
Comment (tecosaur) [reply docs/src/examples.md:62]: It is meant to be impart here, but thanks for checking

File: src/faces.jl [hunk 7]
```diff
@@ -433,12 +433,15 @@ end
...
-Execute `f` with `FACES``.current` temporarily modified by zero or more
-`:name => val` arguments `kv`. `withfaces` is generally used via the
-`withfaces(kv...) do ... end` syntax. A value of `nothing` can be used to
-temporarily unset an face (if if has been set). When `withfaces` returns, the
-original `FACES``.current` has been restored.
+Execute `f` with `FACES``.current` temporarily modified by zero or more `:name
+=> val` arguments `kv`, or `kvpair_itr` which produces `kv`-form values.
+
+`withfaces` is generally used via the `withfaces(kv...) do ... end` syntax. A
+value of `nothing` can be used to temporarily unset an face (if if has been
```
Comment (Seelengrab) [inline src/faces.jl:442]:
```suggestion
value of `nothing` can be used to temporarily unset a face (if it has been
```

Comment (tecosaur) [reply src/faces.jl:442]: Ah yep, ta.

**Comment by tecosaur** [comment]
Ok, I think we're at the point where this can be merged.
